pkg/validator/pvc.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Hewlett Packard Enterprise Development LP
+
+package validator
+
+import (
+	"encoding/json"
+
+	"github.com/bluek8s/kubedirector/pkg/observer"
+	"github.com/bluek8s/kubedirector/pkg/shared"
+	"k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// pvcPatchSpec is used to create the PATCH operation for setting the PVC
+// owner references.
+type pvcPatchSpec struct {
+	Op    string                  `json:"op"`
+	Path  string                  `json:"path"`
+	Value []metav1.OwnerReference `json:"value"`
+}
+
+// mutateOwnerRefs checks PVC labels to see if it is for a kdcluster. If it
+// is, it returns patches to set the ownerReferences to point to that CR.
```

Comment:
👍

Reply:
what about owner ref to statefulset, service etc?

Reply:
We fix those in reconcile instead; see UpdateStatefulSetNonReplicas, UpdateHeadlessService, and UpdatePodService.

Since those owner references are editable (and since we're already in the business of reconciling them) we can do that rather than needing to introduce more webhooks for statefulsets and services.

Reply:
There is kind of an asymmetry here which I'll say a bit more about.

One of the reasons to have a webhook here -- rather than introducing reconciliation handling for PVCs -- is to make sure we handle restored statefulsets that were created by an old version of KD, with the PVC ownerref included in the PVC template of the statefulset. If we don't fix those PVCs by intervening in their creation, then as soon as they are created with an old/stale owner ref they will vanish.

Since we have to have the webhook to handle PVCs generated by those old statefulsets, and the webhook also handles the "normal" case of PVC creation by current-form statefulsets, then we don't need to do any PVC reconciliation.

I guess that does mean that if someone comes along later and manually edits the owner ref on the PVC to remove the kdcluster as owner, we don't forcibly re-instate our kdcluster owner ref there. Might be something we want to think about doing but at least that's not a regression -- that's the way it's always been. I chose not to write a whole new PVC reconciler in this PR just for that purpose. :-)